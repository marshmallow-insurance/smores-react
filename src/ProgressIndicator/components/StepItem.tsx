import { Box } from '../../Box'
import { StepData } from 'ProgressIndicator/types'
import React from 'react'
import styled, { css } from 'styled-components'
import { theme } from '../../theme'
import { Text } from '../../Text'
import { Icon } from '../../Icon'

export interface StepItemProps extends Pick<StepData, 'label'> {
  isCompleted?: boolean
  isLastCompleted?: boolean
  isDisabled: boolean
  isCurrentStep: boolean
  isSimple?: boolean
  onClick: () => void
  stepWidth: string
  isLastItem: boolean
}

export const StepItem = ({
  label,
  isSimple = false,
  isCurrentStep,
  stepWidth,
  onClick,
  isCompleted = false,
  isLastCompleted = false,
  isLastItem = false,
  isDisabled,
}: StepItemProps) => {
  if (isSimple) {
    return (
      <SimpleItem
        $completed={isCompleted}
        $lastCompleted={isLastCompleted}
        width={stepWidth}
        height="12px"
      />
    )
  }

  return (
    <ProgressItem
      flex
      alignItems="flex-start"
      $completed={isCompleted}
      $lastCompleted={isLastCompleted}
      width={stepWidth}
    >
      <ClickableArea
        flex
        data-testid={`step-item-${label}`}
        direction="column"
        alignItems="center"
        onClick={onClick}
        $isDisabled={isDisabled}
      >
        <ProgressIndicator
          $completed={isCompleted}
          $currentStep={isCurrentStep}
          flex
          alignItems="center"
          justifyContent="center"
        >
          {isCompleted && <Icon render="tick" size={16} color="cream" />}
        </ProgressIndicator>
        <StyledText typo="caption">{label}</StyledText>
      </ClickableArea>
      {isCompleted && !isLastItem && <CompletedBar />}
    </ProgressItem>
  )
}

interface StyledComponentProps {
  $completed?: boolean
  $lastCompleted?: boolean
  $currentStep?: boolean
  $isDisabled?: boolean
  $completedStep?: boolean
}

const lastCompleted = css`
  border-radius: 0 100px 100px 0;

  &:first-child {
    border-radius: 100px;
  }
`

const borderRadiusCss = css<StyledComponentProps>`
  &:first-child {
    border-radius: 100px 0 0 100px;
  }

  ${({ $lastCompleted }) => $lastCompleted && lastCompleted}
`

const SimpleItem = styled(Box)<StyledComponentProps>`
  position: relative;
  z-index: 1;
  ${borderRadiusCss}

  background: ${({ $completed }) =>
    $completed ? theme.colors.pistachio : 'none'};
`

const ProgressItem = styled(Box)<StyledComponentProps>`
  position: relative;
  z-index: 1;
  top: 9px;
  left: -10px;

  &:last-child {
    width: auto;
  }
`
const ClickableArea = styled(Box)<StyledComponentProps>`
  position: relative;
  cursor: ${({ $isDisabled }) => ($isDisabled ? 'auto' : 'pointer')};
`

const ProgressIndicator = styled(Box)<StyledComponentProps>`
  border-radius: 50%;
  height: 24px;
  width: 24px;
  position: relative;
  z-index: 1;
  background: ${({ $completed, $currentStep }) =>
    $completed || $currentStep ? theme.colors.pistachio : theme.colors.matcha};
`

const StyledText = styled(Text)`
  margin-top: 2px;
  font-weight: ${theme.font.weight.medium};
  white-space: nowrap;
`

const CompletedBar = styled(Box)`
  position: absolute;
  height: 12px;
  width: 100%;
  top: 7px;
  left: 13px;
  background: ${theme.colors.pistachio};
  z-index: 0;
`
