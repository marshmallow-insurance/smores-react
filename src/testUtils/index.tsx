import React, { ReactElement } from 'react'
import { render, RenderOptions } from '@testing-library/react'
import { SmoresThemeProvider } from '../ThemeProvider/ThemeProvider'

const Providers = ({ children }: { children: React.ReactNode }) => {
  return <SmoresThemeProvider>{children}</SmoresThemeProvider>
}

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, 'wrapper'>,
) => render(ui, { wrapper: Providers, ...options })

export * from '@testing-library/react'
export * from '@testing-library/jest-dom/vitest'
export * from 'jest-styled-components'

export { customRender as render }
