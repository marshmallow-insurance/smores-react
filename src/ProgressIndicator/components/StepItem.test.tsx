import React from 'react'
import { render, screen, fireEvent } from '../../testUtils'
import { StepItem, StepItemProps } from './StepItem'
describe('StepItem', () => {
  const defaultProps: StepItemProps = {
    label: 'Step 1',
    isCurrentStep: false,
    isCompleted: false,
    isLastCompleted: false,
    isDisabled: false,
    isSimple: false,
    onClick: vi.fn(),
    stepWidth: '100px',
    isLastItem: false,
  }

  it('renders the label correctly and does not render tick if notCompleted', () => {
    const { baseElement } = render(<StepItem {...defaultProps} />)

    expect(baseElement).toMatchSnapshot()
    expect(screen.getByText('Step 1')).toBeTruthy()
    expect(screen.queryByTestId('tick-container')).not.toBeTruthy()
  })

  it('renders the completed icon when step is completed', () => {
    const { baseElement } = render(
      <StepItem {...defaultProps} isCompleted={true} />,
    )

    expect(baseElement).toMatchSnapshot()
    expect(screen.getByTestId('tick-container')).toBeTruthy()
  })

  it('calls onClick when clicked', () => {
    render(<StepItem {...defaultProps} />)
    fireEvent.click(screen.getByText('Step 1'))
    expect(defaultProps.onClick).toHaveBeenCalled()
  })

  it('renders a simple item when isSimple is true', () => {
    const { baseElement } = render(
      <StepItem {...defaultProps} isSimple={true} />,
    )
    expect(baseElement).toMatchSnapshot()
    expect(screen.queryByText('Step 1')).not.toBeTruthy()
  })
})
