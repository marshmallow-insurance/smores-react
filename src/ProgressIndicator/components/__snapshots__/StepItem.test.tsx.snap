// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`StepItem > renders a simple item when isSimple is true 1`] = `
<body>
  <div>
    <div
      class="sc-bRKDuR kAMLGE sc-jwTyAe hTtDPN"
    />
  </div>
</body>
`;

exports[`StepItem > renders the completed icon when step is completed 1`] = `
<body>
  <div>
    <div
      class="sc-bRKDuR ioVWEY sc-hjsuWn blvpKV"
    >
      <div
        class="sc-bRKDuR leNTXj sc-jJLAfE ejaShY"
        data-testid="step-item-Step 1"
      >
        <div
          class="sc-bRKDuR fwCHSf sc-hwkwBN ehCFIa"
        >
          <span
            class="sc-bRKDuR bHjpxP sc-fhHczv kxywTO"
            data-testid="tick-container"
          >
            <svg
              fill="none"
              height="100%"
              viewBox="0 0 24 24"
              width="100%"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M20.641 5.232a1.001 1.001 0 00-1.409.128L9.807 16.67l-5.252-3.501a1 1 0 10-1.11 1.664l6 4a1.006 1.006 0 001.324-.193l10-12a1.001 1.001 0 00-.128-1.408z"
                fill="currentColor"
              />
            </svg>
          </span>
        </div>
        <p
          class="sc-bRKDuR bHjpxP sc-dTvVRJ qDeEm sc-kNOymR eiBTXe"
          cursor="inherit"
          title=""
        >
          Step 1
        </p>
      </div>
      <div
        class="sc-bRKDuR bHjpxP sc-dYwGCk kDpsHe"
      />
    </div>
  </div>
</body>
`;

exports[`StepItem > renders the label correctly and does not render tick if notCompleted 1`] = `
<body>
  <div>
    <div
      class="sc-bRKDuR ioVWEY sc-hjsuWn blvpKV"
    >
      <div
        class="sc-bRKDuR leNTXj sc-jJLAfE ejaShY"
        data-testid="step-item-Step 1"
      >
        <div
          class="sc-bRKDuR fwCHSf sc-hwkwBN eWQDzm"
        />
        <p
          class="sc-bRKDuR bHjpxP sc-dTvVRJ qDeEm sc-kNOymR eiBTXe"
          cursor="inherit"
          title=""
        >
          Step 1
        </p>
      </div>
    </div>
  </div>
</body>
`;
